### DISCUSSION TITLE
Apple M1Pro: gfortran not linking

### DISCUSSION THREAD

**Discussion by ChrisNFS** [Homebrew/discussions#2520] [Casks] [answered]
I just installed gcc (brew install gcc; with gfortran included) on a macbook with M1Pro chip running macOS Monterey 12.0.1.
Xtools 13.1 and the command line tools are installed as well.

Unfortunately my fortran codes are no longer linked; after compiling I get this error:

```
_final section layout:
    __TEXT/__text addr=0x10000464C, size=0x0004C2A4, fileOffset=0x0000464C, type=1
    __TEXT/__text_startup addr=0x1000508F0, size=0x0000002C, fileOffset=0x000508F0, type=1
    __TEXT/__stubs addr=0x10005091C, size=0x00000174, fileOffset=0x0005091C, type=29
    __TEXT/__const addr=0x100050A90, size=0x0000CFA9, fileOffset=0x00050A90, type=0
    __TEXT/__cstring addr=0x10005DA40, size=0x00000310, fileOffset=0x0005DA40, type=13
    __TEXT/__eh_frame addr=0x10005DD50, size=0x000062B0, fileOffset=0x0005DD50, type=19
    __DATA_CONST/__got addr=0x100064000, size=0x000000F8, fileOffset=0x00064000, type=30
    __DATA/__data addr=0x100068000, size=0x00000060, fileOffset=0x00068000, type=0
    __DATA/__bss addr=0x100068060, size=0x12B8FC4C0, fileOffset=0x00000000, type=26
    __DATA/__common addr=0x22B964520, size=0x0152DBD4, fileOffset=0x00000000, type=26
ld: ARM64 ADRP out of range (5027655680 max is +/-4GB): from _MAIN__ (0x10000464C) to _dcpu_ (0x22BAC4EA0) in '_MAIN__' from cfem.o for architecture arm64
collect2: error: ld returned 1 exit status
make: *** [solv] Error 1_
```

I have no clue how to handle this kind of problem. Any ideas?

Cheers,
Chris

**Comment by SMillerDev** [comment]
What does `brew config` and `brew doctor` say?

**Reply by ChrisNFS** [reply]
That was quick :)

```
bash-3.2$ brew doctor
Your system is ready to brew.
```

```
bash-3.2$ brew config
HOMEBREW_VERSION: 3.3.5
ORIGIN: https://github.com/Homebrew/brew
HEAD: 354718c40a4e7cc4e5d1d4b03fdce8b992e04e8a
Last commit: 30 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 47e6dc9e7fe7284066ee7159cace7201e7e315fb
Core tap last commit: 3 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.XKZLlhok8J/org.macosforge.xquartz:0
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 13.0.0 build 1300
Git: 2.30.1 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.77.0 => /usr/bin/curl
macOS: 12.0.1-arm64
CLT: 13.1.0.0.1.1633545042
Xcode: 13.1
Rosetta 2: false
bash-3.2$ 
```

**Comment by carlocab** [comment]
I think gfortran has some issues on Monterey, particularly on M1. @fxcoudert would know if this is related.

**Comment by fxcoudert** [comment]
I'm not seeing a generic linking issue with Fortran. This probably depends on details of the code:

```
meau /tmp $ cat a.f90 
subroutine foo(x)
  integer :: x
  print *, sqrt(real(x, kind=8))
end subroutine

program test
  call foo(2)
end program
meau /tmp $ gfortran a.f90 && ./a.out
   1.4142135623730951     
meau /tmp $ gfortran -v              
Using built-in specs.
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/opt/homebrew/Cellar/gcc/11.2.0_2/libexec/gcc/aarch64-apple-darwin20/11/lto-wrapper
Target: aarch64-apple-darwin20
Configured with: ../configure --prefix=/opt/homebrew/Cellar/gcc/11.2.0_2 --libdir=/opt/homebrew/Cellar/gcc/11.2.0_2/lib/gcc/11 --disable-nls --enable-checking=release --with-gcc-major-version-only --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-11 --with-gmp=/opt/homebrew/opt/gmp --with-mpfr=/opt/homebrew/opt/mpfr --with-mpc=/opt/homebrew/opt/libmpc --with-isl=/opt/homebrew/opt/isl --with-zstd=/opt/homebrew/opt/zstd --with-pkgversion='Homebrew GCC 11.2.0_2' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --build=aarch64-apple-darwin20 --with-system-zlib --disable-multilib --with-native-system-header-dir=/usr/include --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.1.0 (Homebrew GCC 11.2.0_2) 
```

**Reply by ChrisNFS** [reply]
You are right - it seems to be not a simple linking problem. At least some memory allocation or stack size things are involved as well. My code is a CFD code, thus needing a lot of memory. Reducing the memory allocation helps, the code compiles, links and runs as expected. Problem is, the (now low) memory allocation allows only to simulate very simple example problems. The same code with much higher memory allocation has been running without trouble on an older 2012 macbook i7 catalina with only 8 gb ram, also using gfortran. Transferring this (unchanged) code to the m1pro with 32gb causes the trouble.

==> In order to check the memory allocation I used a simple code, allocating memory with an array "x1":

```
      program ftest1
      implicit none
      integer i,j,imax, jmax
      parameter (imax = 16 * 1024 * 1024, jmax = 8)
      double precision x1 (imax,jmax)
      do 100 i=1,imax
         do 110 j=1,jmax
            x1 (i,j) = 0.0d0
110      continue
100   continue
      write (*,*) 'array size [mb] = ',dble(imax*jmax*8)/dble(1024**2)
      end
```
 
==> Compiling:

gfortran -Wall ftest1.f -o ftest1.x
ftest1.f:5:37:

    5 |       double precision x1 (imax,jmax)
      |                                     1
Warning: Array 'x1' at (1) is larger than limit set by '-fmax-stack-var-size=', moved from stack to static storage. This makes the procedure unsafe when called recursively, or concurrently from multiple threads. Consider using '-frecursive', or increase the '-fmax-stack-var-size=' limit, or change the code to use an ALLOCATABLE array. [-Wsurprising]

==> Running:

ftest1.x
 array size [mb] =    1024.0000000000000 

==> Ignoring the above compiler warning all seems well. 
==> Now changing the code, using the same parameter but allocating a second array "x2":

```
      program ftest2
      implicit none
      integer i,j,imax, jmax
      parameter (imax = 16 * 1024 * 1024, jmax = 8)
      double precision x1 (imax,jmax)
      double precision x2 (imax,jmax)
      do 100 i=1,imax
         do 110 j=1,jmax
            x1 (i,j) = 0.0d0
            x2 (i,j) = 0.0d0
110      continue
100   continue
      write (*,*) 'array size [mb] = ',dble(imax*jmax*8)/dble(1024**2)
      end
```

==> Compiling:

gfortran -Wall ftest2.f -o ftest2.x
ftest2.f:5:37:

    5 |       double precision x1 (imax,jmax)
      |                                     1
Warning: Array 'x1' at (1) is larger than limit set by '-fmax-stack-var-size=', moved from stack to static storage. This makes the procedure unsafe when called recursively, or concurrently from multiple threads. Consider using '-frecursive', or increase the '-fmax-stack-var-size=' limit, or change the code to use an ALLOCATABLE array. [-Wsurprising]
ftest2.f:6:37:

    6 |       double precision x2 (imax,jmax)
      |                                     1
Warning: Array 'x2' at (1) is larger than limit set by '-fmax-stack-var-size=', moved from stack to static storage. This makes the procedure unsafe when called recursively, or concurrently from multiple threads. Consider using '-frecursive', or increase the '-fmax-stack-var-size=' limit, or change the code to use an ALLOCATABLE array. [-Wsurprising]

==> Same warning as before, only twice given. Now running:

ftest2.x
dyld[4759]: dyld cache '/System/Library/dyld/dyld_shared_cache_arm64e' not loaded: syscall to map cache into shared region failed
dyld[4759]: Library not loaded: /usr/lib/libSystem.B.dylib
  Referenced from: /Users/Grins/Documents/Simulation/Ftest/test1/ftest2.x
  Reason: tried: '/usr/lib/libSystem.B.dylib' (no such file), '/usr/local/lib/libSystem.B.dylib' (no such file)
Abort

==> Now the code allocates not just 1 gb ram as before, but 2 gb. And crashes, hinting a missing library. The compiler warnings were also given with the original CFD code, causing the initial mentioned "linker" trouble. 

On my old macbook i7 catalina with 8 gb ram I received the same warnings when compiling, but BOTH example programs did run as expected without error or crash.

I am not sure how to proceed from this point. Any help is welcome!

**Comment by fxcoudert** [comment, marked as answer]
I can reproduce the issue. I can also reproduce it with a simple C program with static memory:

```
void ftest2 ()
{
  int i, j;
  static double x1[134217728];
  static double x2[134217728];

  for (i = 0; i < 10000; i++)
    for (j = 0; j < 10000; j++)
      x1[i+10000*j] = x2[i+10000*j] = 1;
}

int main (void)
{
  ftest2();
}
```

This is actually even reproducible with Apple's clang compiler:

```
$ clang b.c && ./a.out
dyld[56343]: dyld cache '/System/Library/dyld/dyld_shared_cache_arm64e' not loaded: syscall to map cache into shared region failed
dyld[56343]: Library not loaded: /usr/lib/libSystem.B.dylib
  Referenced from: /private/tmp/a.out
  Reason: tried: '/usr/lib/libSystem.B.dylib' (no such file), '/usr/local/lib/libSystem.B.dylib' (no such file)
zsh: abort      ./a.out
```

So it looks like there is a generic limitation in how much static memory you can map on Apple Silicon (whether from C or Fortran, and regardless of the compiler).

**Reply by fxcoudert** [reply]
Also, please note that the advice given in the warning is working: allocating dynamic memory like this works.

```
      double precision, allocatable :: x1 (:,:), x2(:,:)

      allocate(x1 (imax,jmax))
      allocate(x2 (imax,jmax))
```

**Reply by ChrisNFS** [reply]
Thank you very much for the trouble and the confirmation, that there seems to be a real problem / limitation with Apple Silicon.

Alas my CFD code is of the type "old school". Just plain old FORTRAN (77), no modern gimmicks :) like dynamic memory allocation. It uses a "self made" semi-dynamic memory management by allocating a huge double precision master array and then using many pointers to semi-allocate various integer / double precision smaller arrays within that master array. To integrate real dynamic memory allocation as mentioned by you would be something like "critical surgery" which I rather tend to avoid :)

Can you perhaps suggest any alternative way? Is there a way to increase the stack size mentioned in the warning? If useful, should I contact the apple support? (would you recommend that?)

Again, thank you very much!

**Reply by fxcoudert** [reply]
I don't know if there is a simple solution. Maybe try the Apple developer forums.

**Reply by ChrisNFS** [reply]
I will try that. Thank you for your time and the suggestions!

**Comment by isolated-matrix** [comment]
I know this is an old post, but I'm encountering exactly the same problem (totally unrelated to Homebrew, though). I was just wondering if you found a solution to this? I'm trying to compile a very old fortran program and getting the error:
```
ld: fixup error (kind=arm64_adrp_lo12) at '_buildz_'+0x6A8 from cciJ7bin.o, ADRP out of range, from 0x10001C054 to 0x45E692C60 ('_zeros.2')
final section layout:
    __PAGEZERO           addr=0x000000000, size=0x100000000, fileOffset=0x00000000, fileSize=0x00000000
    __TEXT               addr=0x100000000, size=0x000178000, fileOffset=0x00000000, fileSize=0x00178000
        __text           addr=0x1000006a0, size=0x000158204, fileOffset=0x000006a0 (zerofill=0)
        __stubs          addr=0x1001588a4, size=0x0000001bc, fileOffset=0x001588a4 (zerofill=0)
        __const          addr=0x100158a60, size=0x000017542, fileOffset=0x00158a60 (zerofill=0)
        __cstring        addr=0x10016ffa8, size=0x00000045f, fileOffset=0x0016ffa8 (zerofill=0)
        __eh_frame       addr=0x100170408, size=0x000004128, fileOffset=0x00170408 (zerofill=0)
    __DATA_CONST         addr=0x100178000, size=0x000004000, fileOffset=0x00178000, fileSize=0x00004000
        __got            addr=0x100178000, size=0x000000128, fileOffset=0x00178000 (zerofill=0)
    __DATA               addr=0x10017c000, size=0x35eeac000, fileOffset=0x0017c000, fileSize=0x00004000
        __data           addr=0x10017c000, size=0x000000f54, fileOffset=0x0017c000 (zerofill=0)
        __common         addr=0x10017cf60, size=0x35e515d00, fileOffset=0x00000000 (zerofill=1)
        __bss            addr=0x45e692c60, size=0x000994b80, fileOffset=0x00000000 (zerofill=1)
    __LINKEDIT           addr=0x45f028000, size=0x000008000, fileOffset=0x00180000, fileSize=0x00006f78
collect2: error: ld returned 1 exit status
```

**Reply by fxcoudert** [reply]
You are trying to statically allocate 13 GB of memory, and this is not allowed. Use dynamic allocation.

**Reply by isolated-matrix** [reply]
How can I tell gfortran to compile using dynamic memory allocation? Or is this something at the source code level?

**Reply by SMillerDev** [reply]
Simple search links to: https://www.chriswareham.net/software/FORTRAN77/fortran_malloc.html

**Reply by isolated-matrix** [reply]
Thanks so much for that! I was really hoping I wouldn't have to dig through the source code...
